crates/hir-ty/src/tests/never_type.rs

```diff
@@ -14,6 +14,8 @@ fn test() {
     );
 }
 
+// FIXME(next-solver): The never type fallback implemented in r-a no longer works properly because of
+// `Coerce` predicates. We should reimplement fallback like rustc.
```

Comment:
To expand a bit on this: the never type fallback relies on being able to conclude from the diverging-ness of one infer var to another. It does this, simply, by resolving the root variable. Unfortunately, now this isn't enough: coercion between variables, also an element that needs to be taken into account in the fallback, can be delayed and stored in the form of `Coerce` predicates in the fulfillment context. Worse, unlike root vars, `Coerce` predicates form a graph and not even necessarily a DAG: therefore, some more sophisticated tracking is needed. Instead of complicating this PR even more, I opted to regress never types a bit and leave that as a future work.